### ISSUE TITLE
Get rid of bare "except"s

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#11668]
It looks like quite a few bare `except:` and `except Exception:` lines have crept up throughout the codebase. A few of these are OK (for instance, the ones in `runtests.py` are there to catch and report exceptions in the tests), but most should be replaced with an explicit exception, or rearranged so that the logic does not require checking for an exception. Even places where you really do want to catch all exceptions, `except Exception:` is usually better than `except:`, since the latter also catches things like `KeyboardInterrupt` and `MemoryError`, which you generally really do want to raise to the user. I didn't check all the instances in the code, but I can't think of an example where bare `except:` would be acceptable, excepting the case where there is a `raise` line in the except block. 

I've marked this as easy to fix. It does require a bit of work, however, as you'll need to understand what the code is doing in each instance to know how to fix it (or if it shouldn't be fixed).

**Comment by sidgupta234** [comment]
I wish to take this up.

**Comment by krishnacharya** [comment]
@sidgupta234 Are you still working on this?

**Comment by sidgupta234** [comment]
Nope, I was waiting for his affirmation, which I didn't get.

**Comment by jksuom** [comment]
@sidgupta234  There is no need to wait for an affirmation. You can start working right away on any issue you are interested in.

**Comment by ayodejiige** [comment]
I'm new and would like to take this up if no one is currently working on it.

**Comment by moorepants** [comment]
@ayodejiige please do.

**Comment by nilabja10201992** [comment]
If it is still open, I would like to work with this?

**Comment by OppenheimerAndTheMartians** [comment]
Found some bare excepts...

**Comment by acz13** [comment]
@OppenheimerAndTheMartians fork the repository, commit there, and pull request back here

**Comment by SaiHarsh** [comment]
Is it closed?

**Comment by gxyd** [comment]
**Announcement:** This issue is still _open,_ you can work on it and send in a PR.

**Comment by OppenheimerAndTheMartians** [comment]
I'm sorry, I've been busy with the job search for the past several months. It's taken me far longer than I've ever dreamed, so I haven't been working on the open source projects I want to get involved with.

 

I'll see if I can get around to it soon.

 

Sent: Friday, September 22, 2017 at 11:58 PM

**Comment by SushantGahlot** [comment]
This is still open. Should I go ahead and work on it?

**Comment by OppenheimerAndTheMartians** [comment]
Completely forgot about this... yeah, go on ahead.
 

Sent: Wednesday, July 04, 2018 at 6:58 PM

**Comment by asmeurer** [comment]
Yes, it would likely be better to check for the condition that raises the exception. Or at least catch a more specific exception.

**Comment by mohitacecode** [comment]
yeah,you are right but i think there is no possiblity of any exception.
please correct me if i am wrong

**Comment by asmeurer** [comment]
I think we should add a code quality check for bare `except:`. The few instances where it is acceptable  (basically, when the exception is reraised) can be replaced with `except Exception:` or `except BaseException:`. But for the most part, using `except:` is a serious code smell, that should usually be rewritten to avoid the use of `try` at all, and when that is not the case, it should catch the specific exception.

**Comment by oscarbenjamin** [comment]
Since your comment here appears to be prompted by #15935 I feel I should point out that [as I said there](https://github.com/sympy/sympy/pull/15935#issuecomment-461191411) that case is not a bare except. Actually it is:
```python
    try:
        rf = evalf_table[x.func]
        r = rf(x, prec, options)
    except KeyError:
        try:
            # Fall back to ordinary evalf if possible
```
The problem is that two lines are in the try block but the except is intended only to catch the expected error from the first. This means that exception raising bugs in the second line are hidden. Using a particular exception such as `KeyError` is better than bare except but really the problem here is having too many lines in the try block.

I assume that the original author wrote it that way because
```python
    if x.func in evalf_table:
        rf = evalf_table[x.func]
```
involves twice looking up the `x.func` attribute and it's value from the `evalf_table` dict. A happy medium could be
```python
    rf = evalf_table.get(x.func, None)
    if rf is not None:
```

There are probably many other cases of overly broad try blocks in the code. The `except AttributeError` immediately below the quoted code above looks very much like a bug magnet to me:
```python
def evalf(x, prec, options):
    from sympy import re as re_, im as im_
    try:
        rf = evalf_table[x.func]
        r = rf(x, prec, options)
    except KeyError:
        try:
            # Fall back to ordinary evalf if possible
            if 'subs' in options:
                x = x.subs(evalf_subs(prec, options['subs']))
            xe = x._eval_evalf(prec)
            re, im = xe.as_real_imag()
            if re.has(re_) or im.has(im_):
                raise NotImplementedError
            if re == 0:
                re = None
                reprec = None
            elif re.is_number:
                re = re._to_mpmath(prec, allow_ints=False)._mpf_
                reprec = prec
            else:
                raise NotImplementedError
            if im == 0:
                im = None
                imprec = None
            elif im.is_number:
                im = im._to_mpmath(prec, allow_ints=False)._mpf_
                imprec = prec
            else:
                raise NotImplementedError
            r = re, im, reprec, imprec
        except AttributeError:
            raise NotImplementedError
```
(From [https://github.com/sympy/sympy/blob/master/sympy/core/evalf.py#L1303](https://github.com/sympy/sympy/blob/master/sympy/core/evalf.py#L1303)

**Comment by asmeurer** [comment]
Exactly. Almost every use of a bare `except:` can be replaced with `except SpecificException`, where the specific exception is what the author of the code expected to go wrong. The problem with bare except is it hides any other errors. Even something like misspelling a variable name would raise a NameError and be silenced by the except.

For `KeyError`, you should virtually always use `.get` instead of `except KeyError`, unless you cannot control where place the key lookup is made. Same with `AttributeError`. `except TypeError` and `except ValueError` are also usually suspect, especially since so many different things can raise these exceptions, it's very likely you will end up silencing something beyond what you expected.

In many cases, it is better to just let an exception bubble up instead of guarding against it, for instance, if the exception being raised would indicate a bug or invalid object passed in.
